crunchers/cruncher.h

```diff
@@ -29,58 +29,169 @@
+  using entry_t = ENTRY;
+  using subscriber_callback_t = std::function<void(entry_t&&, idxts_t)>;
 
-  EntryResponse operator()(const entry_t& entry, idxts_t current, idxts_t) {
-    IMPL::OnEvent(entry, current);
-    return EntryResponse::More;
-  }
+  IntermediateSubscriberImpl(subscriber_callback_t callback) : actual_subscriber_(callback) {}
+  virtual ~IntermediateSubscriberImpl() {}
 
   EntryResponse operator()(entry_t&& entry, idxts_t current, idxts_t) {
-    IMPL::OnEvent(std::move(entry), current);
+    actual_subscriber_(std::move(entry), current);
     return EntryResponse::More;
   }
 
   EntryResponse EntryResponseIfNoMorePassTypeFilter() const { return EntryResponse::More; }
   TerminationResponse Terminate() const { return TerminationResponse::Terminate; }
-};
 
-template <typename IMPL>
-using StreamCruncher = current::ss::StreamSubscriber<EntryCruncherImpl<IMPL>, typename IMPL::entry_t>;
+ private:
+  subscriber_callback_t actual_subscriber_;
+};
 
 template <typename ENTRY>
-struct IntermediateSubscriberImpl {
+using IntermediateSubscriber = current::ss::StreamSubscriber<IntermediateSubscriberImpl<ENTRY>, ENTRY>;
+
+template <typename IMPL>
+struct GenericCruncherImpl : public IMPL {
   using EntryResponse = current::ss::EntryResponse;
   using TerminationResponse = current::ss::TerminationResponse;
-  using entry_t = ENTRY;
-  using subscriber_callback_t = std::function<void(entry_t&&, idxts_t)>;
+  using event_t = typename IMPL::event_t;
 
-  IntermediateSubscriberImpl(subscriber_callback_t callback) : actual_subscriber_(callback) {}
-  virtual ~IntermediateSubscriberImpl() {}
+  class Message {
+   public:
+    virtual ~Message() = default;
+    virtual void Handle(IMPL& cruncher) = 0;
+  };
 
-  EntryResponse operator()(entry_t&& entry, idxts_t current, idxts_t) {
-    actual_subscriber_(std::move(entry), current);
+  class EventMessage final : public Message {
+   public:
+    EventMessage(event_t&& e, idxts_t idxts) : event_(std::move(e)), idxts_(idxts) {}
+    EventMessage(const event_t& e, idxts_t idxts) : event_(e), idxts_(idxts) {}
+    void Handle(IMPL& cruncher) override { cruncher.OnEvent(std::move(event_), idxts_); }
+
+   private:
+    event_t event_;
+    const idxts_t idxts_;
+  };
+
+  class RequestMessage final : public Message {
+   public:
+    RequestMessage(Request&& r) : request_(std::move(r)) {}
+    void Handle(IMPL& cruncher) override { cruncher.OnRequest(std::move(request_)); }
+
+   private:
+    Request request_;
+  };
+
+  using mmq_message_t = std::unique_ptr<Message>;
+  using mmq_t = current::mmq::MMQ<mmq_message_t, IntermediateSubscriber<mmq_message_t>, 1024 * 1024>;
+
+  template <typename... ARGS>
+  GenericCruncherImpl(uint16_t port, const std::string& route, ARGS&&... args)
+      : IMPL(std::forward<ARGS>(args)...),
+        port_(port),
+        mmq_subscriber_([this](mmq_message_t&& message, idxts_t) { message->Handle(*this); }),
+        mmq_(mmq_subscriber_, 1024 * 1024) {
```

Comment:
This `1024 * 1024` is probably better as a `long long` template parameter, defaulting to `1024 * 1024`.